capstone-backend/src/main/java/com/google/sps/servlets/SearchServlet.java

```diff
@@ -0,0 +1,181 @@
+      }
+      BufferedReader br = new BufferedReader(new InputStreamReader(
+        (conn.getInputStream())));
+      
+      String output;
+      while ((output = br.readLine()) != null) {
+        fullOutput += output;
+      }
+      conn.disconnect();
+    } catch (MalformedURLException e) {
+      e.printStackTrace();
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+    
+    response.setContentType("application/json;");
+    Gson gson = new GsonBuilder().disableHtmlEscaping().create();
+
+    response.getWriter().println(gson.toJson(convertResponseToVolumeData(fullOutput)));
+  }
+
+  /*
+   * Takes in a response from the Google Books API and converts it to a Collection of VolumeData objects.
+   * @param output the response from the Books API
+   * @return a Collection of the books from the API call, in VolumeData form
+   */
+  @VisibleForTesting
+  static Collection<VolumeData> convertResponseToVolumeData(String output) {
+    ArrayList<VolumeData> volumes = new ArrayList<>();
+
+    // Parse data into json object
+    JsonParser parser = new JsonParser();
+    JsonElement data = parser.parse(output);
+    JsonObject dataInfo = data.getAsJsonObject();
+
+    // Get the items element from the output and convert it into the array of book jsons
+    JsonElement items = dataInfo.get("items");
+    JsonArray itemsInfo = items.getAsJsonArray();
+
+    for (JsonElement book: itemsInfo) {
+      JsonObject bookInfo = book.getAsJsonObject();
+      String id = bookInfo.get("id").getAsString();
+
+      // volumeInfo houses all of the information about the book itself: description, authors, title
+      JsonElement volumeInfo = bookInfo.get("volumeInfo");
+      if (volumeInfo == null) continue;
+      JsonObject volumeInfoObj = volumeInfo.getAsJsonObject();
+
+      JsonElement titleElement = volumeInfoObj.get("title");
+      String title = titleElement != null ? titleElement.getAsString() : "";
+
+      // there may be any number of authors, not guaranteed to only have 1
+      JsonElement authors = volumeInfoObj.get("authors");
+      ArrayList<String> authorNames = new ArrayList<>();
+      if (authors != null) {
+        JsonArray authorsArray = authors.getAsJsonArray();
+        for (JsonElement name : authorsArray) {
+          authorNames.add(name.getAsString());
+        }
+      }
+
+      JsonElement descElement = volumeInfoObj.get("description");
+      String description = descElement != null ? descElement.getAsString() : "";
+
+      JsonElement thumbnailElement = volumeInfoObj.get("imageLinks");
+      String thumbnailLink = thumbnailElement != null ? thumbnailElement
+                                                        .getAsJsonObject()
+                                                        .get("thumbnail")
+                                                        .getAsString()
+                                                        : "";
+      thumbnailLink = thumbnailLink.replace("http", "https");
```

Comment:
Why this line?

Reply:
When deployed on the frontend, the application was trying to load these inside the `img` elements and was loading them insecurely since most of the thumbnails in the API are stored with `http`. The same images can be accessed using the secure `https` so I replaced the URL prefix to load the images securely.